contrib/babelfishpg_tsql/src/hooks.c

```diff
@@ -5193,3 +5212,198 @@ unique_constraint_nulls_ordering(ConstrType constraint_type, SortByDir ordering)
 
 	return SORTBY_NULLS_DEFAULT;
 }
+
+static int32_t
+translate_char_pos(const char* str,
+				   int32_t str_len,
+				   const UChar* str_utf16,
+				   int32_t u16_len, 		/* in 16-bit code units */
+				   int32_t u16_pos,
+				   const char **p_str)
+{
+	UChar32 c;
+	int32_t u16_idx = 0;
+	int32_t out_pos = 0;
+	int32_t u8_offset = 0;
+
+	Assert (GetDatabaseEncoding() == PG_UTF8);
+
+	/* for UTF-8, use ICU macros instead of calling pg_mblen() */
+	while (u16_idx < u16_pos)
+	{
+		U16_NEXT(str_utf16, u16_idx, u16_len, c);
+		U8_NEXT(str, u8_offset, str_len, c);
+		out_pos++;
+	}
+	if (p_str != NULL)
+		*p_str = str + u8_offset;
+
+	return out_pos;
+}
+
+static bool
+pltsql_strpos_non_determinstic(text *t1, text *t2, Oid collid, int *r)
+{
+	pg_locale_t mylocale = 0;
+
+	if (!lc_collate_is_c(collid))
+		mylocale = pg_newlocale_from_collation(collid);
+
+	if (!pg_locale_deterministic(mylocale) && mylocale->provider == 'i')
+	{
+		int32_t len1 = VARSIZE_ANY_EXHDR(t1);
+		int32_t len2 = VARSIZE_ANY_EXHDR(t2);
+		UErrorCode	status = U_ZERO_ERROR;
+		UStringSearch *usearch;
+		UChar *uchar1, *uchar2;
+		int32_t ulen1, ulen2;
+		int32_t u16_pos, u8_pos = 0;
+
+		ulen1 = icu_to_uchar(&uchar1, VARDATA_ANY(t1), len1);
```

Comment:
As we are invoking ICU APIs, let's put safety check:
#ifdef USE_ICU
#else
	ereport(ERROR,
			(errcode(ERRCODE_EXTERNAL_ROUTINE_EXCEPTION),
				errmsg("ICU library is required to be installed in order to use the function remove_accents_internal")));
	PG_RETURN_NULL();
#endif
Ref: https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/BABEL_4_X_DEV/contrib/babelfishpg_tsql/src/collation.c#L482

Reply:
done